jwt/src/main/java/io/scalecube/security/AsyncJwtAuthenticator.java

```diff
@@ -0,0 +1,47 @@
+package io.scalecube.security;
+
+import io.jsonwebtoken.Claims;
+import io.jsonwebtoken.Header;
+import io.jsonwebtoken.Jws;
+import io.jsonwebtoken.Jwt;
+import io.jsonwebtoken.JwtParser;
+import io.jsonwebtoken.Jwts;
+import reactor.core.publisher.Mono;
+
+@SuppressWarnings("rawtypes")
+public final class AsyncJwtAuthenticator implements JwtAuthenticator {
+
+  private final AsyncJwtKeyResolver jwtKeyResolver;
+  private static final JwtParser jwtParser = Jwts.parser();
+
+  public AsyncJwtAuthenticator(AsyncJwtKeyResolver jwtKeyResolver) {
+    this.jwtKeyResolver = jwtKeyResolver;
+  }
+
+  @Override
+  public Mono<Profile> authenticate(String token) {
+    return Mono.just(token)
+        .map(AsyncJwtAuthenticator::jwtFromJwsToken)
+        .map(AsyncJwtAuthenticator::parseClaimsFromJwt)
+        .map(AsyncJwtAuthenticator::keyIdFromClaims)
+        .flatMap(jwtKeyResolver::resolve)
+        .map(Jwts.parser()::setSigningKey)
+        .map(jwsParser -> jwsParser.parseClaimsJws(token))
```

Comment:
A token is parsing twice - on line 25 and on line 29, which is not good because of performance reasons.

Since you have already split token, try to use `SignatureValidator` or `JwtSignatureValidator`  of `jjwt` directly to verify the signature.

Reply:
please review on a new branch `redesign-auth-refactor`